libfoundation/src/foundation-string.cpp

```diff
@@ -1111,28 +1442,46 @@ uindex_t MCStringGetChars(MCStringRef self, MCRange p_range, unichar_t *p_chars)
 	return t_count;
 }
 
-uindex_t MCStringGetNativeChars(MCStringRef self, MCRange p_range, char_t *p_chars)
+uindex_t MCStringGetNativeCharsWithReplacement(MCStringRef self, MCRange p_range, char_t *p_replacement, char_t *p_chars)
 {
     if (__MCStringIsIndirect(self))
         self = self -> string;
     
-	uindex_t t_count;
-	t_count = 0;
-
-	for(uindex_t i = p_range . offset; i < p_range . offset + p_range . length; i++)
-	{
-		if (i >= self -> char_count)
-			break;
+    uindex_t t_count;
+    t_count = 0;
+    
+    for(uindex_t i = p_range . offset; i < p_range . offset + p_range . length; i++)
+    {
+        if (i >= self -> char_count)
+            break;
         
+        char_t t_native_char;
         if (MCStringIsNative(self))
-            p_chars[i - p_range . offset] = self -> native_chars[i];
+            t_native_char = self -> native_chars[i];
         else
-            p_chars[i - p_range . offset] = MCStrCharMapToNative(self -> chars[i]);
-
-		t_count += 1;
-	}
+        {
+            if (!MCUnicodeCharMapToNative(self -> chars[i], t_native_char))
+            {
+                if (p_replacement)
```

Comment:
Having `p_replacement` passed in as a pointer suggests that `p_replacement` could be a multi-byte sequence (e.g. `BAD`).

If `*p_replacement` is `\0`, consider abandoning the loop early.